src/finite_elasticity_routines.f90

```diff
@@ -8492,6 +9545,441 @@ END SUBROUTINE FINITE_ELASTICITY_PRE_SOLVE
+                CALL MATRIX_PRODUCT(F_e_T,F_e,C_e,ERR,ERROR,*999)
+    
+                !Odgen law - 3 terms. Material Parameters C = [mu(1) mu(2) mu(3) alpha(1) alpha(2) alpha(3) mu_0]
+                !CALL Eigenvalue(C_e,EVALUES,ERR,ERROR,*999)
+                CALL DSYEV('V','U',3,C_e,3,EVALUES,WORK,-1,ERR)
+                IF(ERR.NE.0) WRITE(*,*) "Error in Eigenvalue computation"
+                LWORK=MIN(LWMAX,INT(WORK(1)))
+                CALL DSYEV('V','U',3,C_e,3,EVALUES,WORK,LWORK,ERR)
+                IF(ERR.NE.0) WRITE(*,*) "Error in Eigenvalue computation"
+                EVECTOR_1=C_e(:,1)
+                EVECTOR_2=C_e(:,2)
+                EVECTOR_3=C_e(:,3)
+
+                DO i=1,3
+                  DO j=1,3
+                    EMATRIX_1(i,j)=EVECTOR_1(i)*EVECTOR_1(j)
+                    EMATRIX_2(i,j)=EVECTOR_2(i)*EVECTOR_2(j)
+                    EMATRIX_3(i,j)=EVECTOR_3(i)*EVECTOR_3(j)
+                  END DO
+                END DO
+
+                CALL MATRIX_PRODUCT(F_a_inv,EMATRIX_1,N1,ERR,ERROR,*999)
+                CALL MATRIX_PRODUCT(N1,F_a_inv,N1,ERR,ERROR,*999) ! F_a_inv=F_a_inv_T
+                CALL MATRIX_PRODUCT(F_a_inv,EMATRIX_2,N2,ERR,ERROR,*999)
+                CALL MATRIX_PRODUCT(N2,F_a_inv,N2,ERR,ERROR,*999) ! F_a_inv=F_a_inv_T
+                CALL MATRIX_PRODUCT(F_a_inv,EMATRIX_3,N3,ERR,ERROR,*999)
+                CALL MATRIX_PRODUCT(N3,F_a_inv,N3,ERR,ERROR,*999) ! F_a_inv=F_a_inv_T 
+
+                FREE_ENERGY_0=0.0_DP
+                DO i=1,3
+                  FREE_ENERGY_0=FREE_ENERGY_0+C(i)/C(i+3)*( &
+                    & EVALUES(1)**(C(i+3)/2.0_DP)+ &
+                    & EVALUES(2)**(C(i+3)/2.0_DP)+ &
+                    & EVALUES(3)**(C(i+3)/2.0_DP)-3.0_DP)
+                END DO
+                FREE_ENERGY_0=C(7)*FREE_ENERGY_0
+
+                FREE_ENERGY=FREE_ENERGY_0
+
+                VALUE=XB_ENERGY_PER_VOLUME-(FREE_ENERGY-FREE_ENERGY_0)
+
+                !tolerance for Newton's method
+                TOL=0.00001_DP
+                !tolerance for the bisection method as preconditioner. Since Newton's method does not converge, we only use the bisection method here
+                TOL1=TOL 
+                UP=lambda_a
+                LOW=0.001_DP
+
+                DO WHILE (ABS(VALUE).GE.TOL)
+
+                  !bisection method
+                  IF (ABS(VALUE).GE.TOL1) THEN
+                    lambda_a=UP-(UP-LOW)/2.0_DP
+
+                    F_a_inv=0.0_DP
+                    F_a_inv(1,1)=1.0_DP/lambda_a
+                    F_a_inv(2,2)=1.0_DP
+                    F_a_inv(3,3)=1.0_DP
+
+                    CALL MATRIX_PRODUCT(DZDNU,F_a_inv,F_e,ERR,ERROR,*999)
+                    CALL MATRIX_TRANSPOSE(F_e,F_e_T,ERR,ERROR,*999)
+                    CALL MATRIX_PRODUCT(F_e_T,F_e,C_e,ERR,ERROR,*999)
+
+                    CALL DSYEV('V','U',3,C_e,3,EVALUES,WORK,-1,ERR)
+                    IF(ERR.NE.0) WRITE(*,*) "Error in Eigenvalue computation"
```

Comment:
WRITE(*,*) should not be used (and throughout). A proper FlagError should be used here.